### DISCUSSION TITLE
Basic TcpListener for callback

### DISCUSSION THREAD

**Discussion by ShayBox** [ramsayleung/rspotify#365] [Ideas]
Why copy/paste the URL or Code into the terminal, why not bind a simple tcp listener for the callback url?  
Add a parameter `addr: Option<SocketAddr>` to `prompt_for_token(&url, "127.0.0.1:8888");`  
If provided, bind a listener, [here's an example](https://doc.rust-lang.org/book/ch20-01-single-threaded.html) using just std::net and possibly std::io  
If it fails to bind or return a valid code within X requests, fallback to stdin input, or alternatively, create duplicate functions for this purpose and leave the existing ones as-is.

**Comment by ShayBox** [comment]
I created my concept <https://github.com/ShayBoxs-Forks/rspotify/commit/362b1ebe8e69712b9f2994f8bd95a2bae199e63a?diff=unified>

**Comment by ramsayleung** [comment]
Great point!

It's much better to ask the developer to copy/paste the URL into the terminal. You could make a PR to add this `TCPListener` support.

**Reply by ShayBox** [reply]
It's much better to handle the callback from Spotify as-intended by the oauth protocol and developers.  
I opted to use tiny_http in my project since I couldn't figure out how to respond with html using TcpListener, though the code works without a response, I wanted the tab to close with javascript.  

Here's what I ended up with if anyone wants to use it in their project, or implement it into rSpotify as a feature flag, as copy/pasting the url from an error page is not user-friendly.  
https://github.com/ShayBox/Spotify-VRChat-OSC/blob/master/src/main.rs

**Reply by marioortizmanero** [reply]
The `prompt_for_token` function is just meant to be a simple implementation. There are many ways to handle the callback for production-oriented use-cases, including servers like Actix. I would actually expect users to implement their own function, and only use `prompt_for_token` when prototyping. Unfortunately, your links seem to be broken now, so I can't take a look.

**Reply by ShayBox** [reply]
Whoops, I changed my repo name I guess the link invalidated
But yeah basically what you said, but I used tiny_http as this is running on a local system and handles a single request  
https://github.com/ShayBox/VRC-OSC/blob/master/plugin-spotify/src/lib.rs#L211

**Comment by gdesmott** [comment]
See also  `librespot-oauth` doing something similar: https://github.com/librespot-org/librespot/blob/dev/oauth/src/lib.rs

**Comment by ramsayleung** [comment]
cool, I am migrating this feature to RSpotify.

**Comment by ramsayleung** [comment]
this feature has been implemented in this PR: https://github.com/ramsayleung/rspotify/pull/513
